crates/ruff_linter/src/suppression.rs

```diff
@@ -138,13 +148,47 @@ impl Suppressions {
             return false;
         };
 
-        for suppression in &self.valid {
+        for suppression in &mut self.valid {
             if *code == suppression.code.as_str() && suppression.range.contains_range(range) {
+                suppression.used = true;
                 return true;
             }
         }
         false
     }
+
+    pub(crate) fn generate_diagnostics(&self, context: &LintContext, locator: &Locator) {
+        if context.is_rule_enabled(Rule::UnusedSuppression) {
+            for suppression in &self.valid {
+                if !suppression.used {
+                    for comment in &suppression.comments {
+                        let edit = if comment.codes.len() == 1 {
+                            delete_comment(comment.range, locator)
+                        } else {
+                            let code_index = comment
+                                .codes
+                                .iter()
+                                .position(|range| locator.slice(range) == suppression.code)
+                                .unwrap();
+                            let code_range = if code_index < (comment.codes.len() - 1) {
+                                TextRange::new(
+                                    comment.codes[code_index].start(),
+                                    comment.codes[code_index + 1].start(),
+                                )
+                            } else {
+                                comment.codes[code_index]
+                            };
+                            Edit::range_deletion(code_range)
+                        };
```

Comment:
I think the logic here only works now because there are always at most 2 comments. This loop would, otherwise, delete the `+1` comment twice. 

I'd be leaning towards splitting `comments` into two fields: `disable_comment` (could also be an `ignore` in the future), `enable_comment: Option<...>`. This way, making the assumption here that we always have at most two comments is safe (and it also removes the need to find the comment again.)